neurom/core/types.py

```diff
@@ -110,43 +118,38 @@ def __ne__(self, other):
         """Not equal operator."""
         return not self == other
 
-    @classmethod
-    def from_list(cls, values):
-        """Create a NeuriteType from a list of sub types."""
-        value = 0
-        for num, i in enumerate(values[::-1]):
-            new_value = cls._BASE**num * int(i)
-            value += new_value
-        obj = cls(value)
-        return obj
-
-    def to_list(self):
-        """Transform a NeuriteType or a positive integer in a list of sub types."""
-        digits = []
-        tmp_value = int(self)
-        while tmp_value:
-            digits.append(int(tmp_value % self._BASE))
-            tmp_value //= self._BASE
-        if not digits:
-            digits = [0]
-        res = digits[::-1]
-        if _ALL_SUBTYPE in res and len(res) > 1:
-            raise NeuroMError(
-                f"A subtype containing the value {_ALL_SUBTYPE} must contain only one element "
-                f"(current elements: {res})."
-            )
-        return res
-
     @property
     def root_type(self):
         """Get the root type of a composite type."""
-        return self.to_list()[0]
+        return self._subtypes[0]
 
     def __reduce_ex__(self, *args, **kwargs):
         """This is just to ensure the type is recognized as picklable by the Enum class."""
         return super().__reduce_ex__(*args, **kwargs)
 
 
+def _ids_to_index(ids, base):
+    """Combine ids on a square grid with side 'base' into a single linear index."""
+    if len(ids) == 1:
+        return ids[0]
+    return int(np.ravel_multi_index(ids, (base,) * len(ids)))
+
+
+def _index_to_ids(index, base):
+    """Convert a linear index into ids on a square grid with side 'base'."""
+    # find number of integers in linear index
+    n_digits = math.ceil(len(str(index)) / (len(str(base)) - 1))
+
+    int_types = np.unravel_index(index, shape=(base,) * n_digits)
+
+    if _ALL_SUBTYPE in int_types and len(int_types) > 1:
+        raise NeuroMError(
+            f"A subtype containing the value {_ALL_SUBTYPE} must contain only one element "
+            f"(current elements: {int_types})."
+        )
+    return int_types
```

Comment:
I think these should be static methods of `SubtypeCollection` since they are only related to this class.

Reply:
I don't have a strong preference although I always prefer to decouple. But in this case you are right that they are not used anywhere else. I will move them inside as static methods.